flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java

```diff
@@ -137,9 +211,46 @@ void onConsumedPartition(ResultPartition partition) {
 
 		// Release the partition if it was successfully removed
 		if (partition == previous) {
-			partition.release();
+			// move to cache if cachable
+			updateIntermediateResultPartitionCache(partition);
 
-			LOG.debug("Released {}.", partition);
+			LOG.debug("Cached {}.", partition);
+		}
+	}
+
+	/**
+	 * Triggered by @link{NetworkBufferPool} when network buffers should be freed
+	 * @param requiredBuffers The number of buffers that should be cleared.
+	 */
+	public boolean releaseLeastRecentlyUsedCachedPartitions (int requiredBuffers) {
+		synchronized (cachedResultPartitions) {
+			// make a list of ResultPartitions to release
+			List<ResultPartition> toBeReleased = new ArrayList<ResultPartition>();
+			int numBuffersToBeFreed = 0;
+
+			// traverse from least recently used cached ResultPartition
+			for (Map.Entry<IntermediateResultPartitionID, ResultPartition> entry : cachedResultPartitions.entrySet()) {
+				ResultPartition cachedResult = entry.getValue();
+
+				synchronized (registeredPartitions) {
+					if (!registeredPartitions.containsValue(cachedResult)) {
+						if (numBuffersToBeFreed < requiredBuffers) {
+							toBeReleased.add(cachedResult);
+							numBuffersToBeFreed += cachedResult.getTotalNumberOfBuffers();
```

Comment:
Sorry for the bad naming, but this is the total number of produced buffers. We need the number of buffers, which are currently in memory.